packages/muelu/research/regionMG/src/SetupRegionSmoothers_def.hpp

```diff
@@ -94,92 +94,75 @@ std::map<std::string, int> getListOfValidSmootherTypes()
 
-  // Get max number of regions per proc
-  const int maxRegPerProc = regionGrpMats.size();
+  RCP<Vector> diagReg = VectorFactory::Build(revisedRowMapPerGrp, true);
 
-  // extract diagonal from region matrices, recover true diagonal values, invert diagonal
-
-  Array<RCP<Vector> > diagReg(maxRegPerProc);
-  createRegionalVector(diagReg, revisedRowMapPerGrp);
-
-  for (int j = 0; j < maxRegPerProc; j++) {
-    // extract inverse of diagonal from matrix
-    diagReg[j] = VectorFactory::Build(regionGrpMats[j]->getRowMap(), true);
-    regionGrpMats[j]->getLocalDiagCopy(*diagReg[j]);
-  }
+  // extract inverse of diagonal from matrix
+  diagReg = VectorFactory::Build(regionGrpMats->getRowMap(), true);
+  regionGrpMats->getLocalDiagCopy(*diagReg);
 
   sumInterfaceValues(diagReg, revisedRowMapPerGrp, rowImportPerGrp);
 
-  for (int j = 0; j < maxRegPerProc; j++) {
-    diagReg[j]->reciprocal(*diagReg[j]);
-  }
+  diagReg->reciprocal(*diagReg);
 
-  params->set<Teuchos::Array<RCP<Vector> > >("smoothers: inverse diagonal", diagReg);
+  params->set<RCP<Vector> >("smoothers: inverse diagonal", diagReg);
 }
 
+
 /*! \brief Do Jacobi smoothing
  *
  *  Perform Jacobi smoothing in the region layout using the true diagonal value
  *  recovered from the splitted matrix.
  */
 template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
 void jacobiIterate(RCP<Teuchos::ParameterList> smootherParams,
-                   Array<RCP<Xpetra::Vector<Scalar, LocalOrdinal, GlobalOrdinal, Node> > >& regX, // left-hand side (or solution)
-                   const Array<RCP<Xpetra::Vector<Scalar, LocalOrdinal, GlobalOrdinal, Node> > > regB, // right-hand side (or residual)
-                   const std::vector<RCP<Xpetra::Matrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> > > regionGrpMats, // matrices in true region layout
-                   const std::vector<RCP<Xpetra::Map<LocalOrdinal, GlobalOrdinal, Node> > > revisedRowMapPerGrp, ///< revised row maps in region layout [in] (actually extracted from regionGrpMats)
-                   const std::vector<RCP<Xpetra::Import<LocalOrdinal, GlobalOrdinal, Node> > > rowImportPerGrp,///< row importer in region layout [in]
+                   RCP<Xpetra::Vector<Scalar, LocalOrdinal, GlobalOrdinal, Node> >& regX, // left-hand side (or solution)
+                   const RCP<Xpetra::Vector<Scalar, LocalOrdinal, GlobalOrdinal, Node> > regB, // right-hand side (or residual)
+                   const RCP<Xpetra::Matrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> > regionGrpMats, // matrices in true region layout
+                   const RCP<Xpetra::Map<LocalOrdinal, GlobalOrdinal, Node> > revisedRowMapPerGrp, ///< revised row maps in region layout [in] (actually extracted from regionGrpMats)
+                   const RCP<Xpetra::Import<LocalOrdinal, GlobalOrdinal, Node> > rowImportPerGrp,///< row importer in region layout [in]
```

Comment:
The removed `std::vector` was related to the concept of "groups". Hence, this could be renamed to 
```suggestion
                   const RCP<Xpetra::Matrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> > regionMats, // matrices in true region layout
                   const RCP<Xpetra::Map<LocalOrdinal, GlobalOrdinal, Node> > revisedRowMap, ///< revised row maps in region layout [in] (actually extracted from regionGrpMats)
                   const RCP<Xpetra::Import<LocalOrdinal, GlobalOrdinal, Node> > rowImport,///< row importer in region layout [in]
```